src/resources.rs

```diff
@@ -2,6 +2,106 @@
 // SPDX-License-Identifier: Apache-2.0
 
 //! Structs to manage device resources.
+//!
+//! The high level flow of resource management among the VMM, the device manager, and the device
+//! is as below:
+//! 1) the VMM creates a new device object.
+//! 2) the VMM asks the new device object for its resource requirements.
+//! 3) the VMM allocates resources for the device object according to resource requirements.
+//! 4) the VMM passes the allocated resources to the device object.
+//! 5) the VMM registers the new device onto corresponding device managers according the allocated
+//!    resources.
+
+use std::{u16, u32, u64};
+
+/// Enumeration describing a device's resource requirements.
+pub enum DeviceResourceRequest {
```

Comment:
I found there is a `DeviceResources` in previous commit, which is used to describe a specific device resource request. Is `DeviceResourceRequest` global resources of a VM, that is to say, it acts the same as vm-allocator crate before? If not, what's the difference between DeviceResources` and `DeviceResourceRequest` ?

Reply:
The vmm asks each device for resource it needs, and the devices returns an array of DeviceResourceRequest to describe resource type and allocation constraints.
Then the VMM passes an array of allocated resource to the device, which is type of DeviceResources.

Reply:
Thanks. I'm thinking whether we could merge `DeviceResourceRequest ` and `DeviceResources` into one type?

Reply:
The have different roles. The DeviceResourceRequest focuses on constraints, and DeviceResource focuses on resources.